File: src/main/java/yuuto/enhancedinventories/item/ItemFunctionUpgrade.java
package yuuto.enhancedinventories.item;

import java.util.List;

import net.minecraft.item.ItemStack;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraftforge.common.util.ForgeDirection;
import yuuto.enhancedinventories.EnhancedInventories;
import yuuto.enhancedinventories.tile.TileConnectiveInventory;
import yuuto.yuutolib.item.ModItemMulti;

public class ItemFunctionUpgrade extends ModItemMulti{

	public ItemFunctionUpgrade() {
		super(EnhancedInventories.tab, "EnhancedInventories", "functionUpgrade", ".base", ".hopper", ".redstone");
	}
	
	@Override
    public boolean onItemUseFirst(ItemStack stack, EntityPlayer player, World world, int X, int Y, int Z, int side, float hitX, float hitY, float hitZ)
    {
		if(world.isRemote || stack.getItemDamage() == 0 || !player.isSneaking())
			return false;
		TileEntity tile = world.getTileEntity(X, Y, Z);
		if(tile == null || !(tile instanceof TileConnectiveInventory))
			return false;
		TileConnectiveInventory oldTile = (TileConnectiveInventory) tile;
		if(!oldTile.canUpgrade(stack))
			return false;
		TileConnectiveInventory newTile = oldTile.getUpgradeTile(stack);
		if(newTile == null || newTile == oldTile ||!canUpgrade(oldTile, newTile, world, X, Y, Z))
			return false;
		newTile.setContents(oldTile.getContents(), true);
		oldTile.invalidate();
		world.setTileEntity(X, Y, Z, newTile);
		world.setBlockMetadataWithNotify(X, Y, Z, world.getBlockMetadata(X, Y, Z), 3);
		oldTile.invalidate();
		newTile.disconnect();
		newTile.checkConnections();
		newTile.markDirty(true);
		if(!player.capabilities.isCreativeMode){
			stack.stackSize--;
			player.inventoryContainer.detectAndSendChanges();
		}
		return true;
    }
	
	
	public boolean canUpgrade(TileConnectiveInventory oldTile, TileConnectiveInventory newTile, World world, int x, int y, int z){
		List<ForgeDirection> dirs = newTile.getValidConnectionSides();
    	int chests = 0;
    	for(ForgeDirection dir : dirs){
    		TileEntity tile = world.getTileEntity(x+dir.offsetX, y+dir.offsetY, z+dir.offsetZ);
    		if(tile == null || !(tile instanceof TileConnectiveInventory))
    			continue;
    		TileConnectiveInventory inv = (TileConnectiveInventory)tile;
    		if(!inv.isValidForConnection(newTile))
    			continue;
    		if(inv.getPartner() != null)
    			return false;
    		chests++;
    	}
    	if(chests > 1)
    		return false;
    	return true;
	}

}

File: src/main/scala/yuuto/enhancedinventories/item/ItemFunctionUpgrade.scala
package yuuto.enhancedinventories.item

import java.util.List
import yuuto.enhancedinventories.tile.traits.TUpgradeable
import yuuto.yuutolib.item.traits.TItemMultiBase
import net.minecraft.tileentity.TileEntity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.world.World
import yuuto.enhancedinventories.item.base.ItemBaseEI


class ItemFunctionUpgrade(name:String) extends ItemBaseEI(name) with TItemMultiBase{
  this.hasSubtypes = true;
  
  override def onItemUse(stack:ItemStack, player:EntityPlayer, world:World, x:Int, y:Int, z:Int, side:Int, hitX:Float, hitY:Float, hitZ:Float):Boolean={
    if(!player.isSneaking() || world.isRemote || stack.getItemDamage() == 0)
      return false;
    val tile:TileEntity = world.getTileEntity(x, y, z);
    if(tile == null || !tile.isInstanceOf[TUpgradeable])
      return false;
    val chest:TUpgradeable = tile.asInstanceOf[TUpgradeable];
    if(!chest.isUpgradeValid(stack, player)){
      if(stack.getItemDamage() == 3)
      return false;
    }
    if(!chest.addUpgrade(stack, player)){
      if(stack.getItemDamage() == 3)
      return false;
    }
    if(player.capabilities.isCreativeMode)
      return true;
    stack.stackSize -=1;
    player.inventoryContainer.detectAndSendChanges();
    return true;
  }

}